### PR TITLE
feat: add the new private-chat command

### LINKED ISSUE #13: ?private-chat command
```
?private-chat @JacobMGEvans @marcosvega91 
```

This would create a channel that expires after 10 minutes of inactivity (and a max of 1 hour) that's only accessible to the command-er and the people mentioned.

Would require a command function as well as a cleanup.

I'm mostly making not of this myself because there's still not any easy way for people to contribute to this project without making their own server to test things out.

### PR DESCRIPTION
**Pull request by marcosvega91**
fix #13 

**What**: I have added the new awesome command. I'm not pretty sure about two things 
1) Texts in general 😜. My english is not the best
2) The naming convention used for the channel

**Why**: Because `discord` doesn't have threads. This new command is not like a thread but can help members to move on a new chat when discussing on something

**How**: The new command is available as ?private-chat.
Members can create new channels using this command, but can be only one channel active for the same members.
The channel will be created under a category called `PRIVATE CHAT`. 
After 10 minutes of inactivity or after 1 hour the chat will be deleted. There is also a warning message 5 minutes before delete.

About testing, I think that testing the bot should be pretty simple. For the `cleanup` should be a little bit complicate 

**Checklist**:

- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged

### CODE REVIEW

File: src/private-chat/cleanup.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      if (timeSinceChannelCreation > maxExistingTime) {
+        reason = eolReason
+      } else {
+        reason = inactivityReason
+      }
+      await send(
+        `
+    This channel is getting deleted for the following reason: ${reason}
+    
+    Goodbye 👋
```
Comment (kentcdodds) [inline src/private-chat/cleanup.js]:
```suggestion
This channel is getting deleted for the following reason: ${reason}

Goodbye 👋
```

File: src/private-chat/cleanup.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    ) {
+      let reason
+      if (timeSinceChannelCreation > maxExistingTime - warningStep) {
+        reason = eolReason
+      } else {
+        reason = inactivityReason
+      }
+      await send(
+        `
+    This channel will be deleted in 5 minutes for the following reason: ${reason}
```
Comment (kentcdodds) [inline src/private-chat/cleanup.js]:
```suggestion
This channel will be deleted in 5 minutes for the following reason: ${reason}
```

File: src/private-chat/cleanup.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+async function cleanup(guild) {
+  const categoryPrivateChat = guild.channels.cache.find(
+    ({name, type}) =>
+      type === 'category' && name.toLowerCase().includes('private chat'),
+  )
+
+  const warningStep = 1000 * 60 * 5
+  const maxExistingTime = 1000 * 60 * 60
+  const maxInactiveTime = 1000 * 60 * 10
```
Comment (JacobMGEvans) [inline src/private-chat/cleanup.js:11]: Not sure if this is too long or too short, 10mins seems a bit long to me.

File: src/private-chat/cleanup.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
+const {getSend, sleep} = require('../utils')
+const warnedChannels = new Set()
```
Comment (kentcdodds) [inline src/private-chat/cleanup.js]: If the server gets restarted, this set will disappear. So instead maybe you could check the messages in the channel and see whether they've been warned already.

File: src/private-chat/cleanup.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      await send(
+        `
+    This channel is getting deleted for the following reason: ${reason}
+    
+    Goodbye 👋
+        `.trim(),
+      )
+
+      // Give just a while for the users to understand that the channel will be deleted soon
+      await sleep(3000)
```
Comment (kentcdodds) [inline src/private-chat/cleanup.js]: Let's give them more notice:

```suggestion
      await sleep(10000)
```

For this to work though we should see whether there's already a message telling them it's going to be deleted and if there is, then we don't bother telling them again (unless that message is older than 10 seconds, in which case we'd just delete the channel).

File: src/private-chat/cleanup.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    channel =>
+      channel.type === 'text' &&
+      channel.parentID === categoryPrivateChat.id &&
+      channel.name.includes('-private-') &&
+      !channel.deleted,
+  )
+
+  allActivePrivateChannels.forEach(async channel => {
+    const channelCreateDate = channel.createdAt
+    const lastMessageDate = channel.lastMessage?.createdAt ?? channelCreateDate
```
Comment (JacobMGEvans) [inline src/private-chat/cleanup.js:26]: Are there instances when the `createdAt` property is not on `lastMessage`? 

I assume the possibility is covering when the channel is opened however we could have the bot have a welcome message or creation message in the chat as well

File: src/private-chat/cleanup.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
+const {getSend, sleep} = require('../utils')
+const warnedInactiveChannels = new Set()
+const warnedEOLChannels = new Set()
```
Comment (kentcdodds) [inline src/private-chat/cleanup.js]: Unfortunately if the server is restarted then we lose this state. So instead we should determine the warning status based on whether the bot has sent the message or not.

File: src/commands/command-fns/private-chat.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        ...membersPermissions,
+      ],
+    },
+  )
+
+  const send = getSend(channel)
+
+  await send(
+    `
+Hello all 👋
```
Comment (kentcdodds) [inline src/commands/command-fns/private-chat.js]: Let's mention all the users so they get a notification.

File: src/commands/command-fns/private-chat.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+          type: 'role',
+          id: everyoneRole.id,
+          deny: allPermissions,
+        },
+        ...membersPermissions,
+      ],
+    },
+  )
+
+  const send = getSend(channel)
```
Comment (kentcdodds) [inline src/commands/command-fns/private-chat.js]: `getSend` is only really useful if we send multiple messages and want to wait between them. We can just use `channel.send` here.

File: src/commands/command-fns/private-chat.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+      .map(member => member.id)
+
+    return (
+      chatMembers.length === mentionedMembers.length &&
+      mentionedMembers.every(member => chatMembers.indexOf(member.id) !== -1)
+    )
+  })
+
+  if (existingChat) {
+    return message.channel.send(`There is already a chat for the same members`)
```
Comment (kentcdodds) [inline src/commands/command-fns/private-chat.js]: Nice feature! Could we reference it?

```suggestion
    return message.channel.send(`There is already a chat for the same members ${existingChat}`)
```

**Review by kentcdodds — COMMENTED** [review]
This is looking amazing 👏

File: src/private-chat/cleanup.js (hunk 6 above)
Comment (marcosvega91) [reply src/private-chat/cleanup.js:26]: Yes lastMessage should always be there. Because the bot will be the first message

Comment (kentcdodds) [reply src/private-chat/cleanup.js:26]: I don't mind being overly cautious here... Remember that the bot can get restarted any time so weird things can happen.

**Comment by marcosvega91** [comment]
I have made the changes. If server crashed or will be restarted interrupting the sleep, after two minute from the first deletion I'll delete the channel again.

File: src/private-chat/cleanup.js (hunk 3 above)
Comment (marcosvega91) [reply src/private-chat/cleanup.js:11]: maybe 5 can be better or 8 ? mmm

Comment (JacobMGEvans) [reply src/private-chat/cleanup.js:11]: I was personally thinking 5 mins, but I don't know how the private convos will be primarily used.

**Review by JacobMGEvans — APPROVED** [review]
Other than my opinion on the time for inactivity, this is super solid! Great work 🎉 🎊

File: src/private-chat/cleanup.js (hunk 3 above)
Comment (marcosvega91) [reply src/private-chat/cleanup.js:11]: of course we can change if we understand that a different value is better

File: src/private-chat/cleanup.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
+const {getSend, sleep} = require('../utils')
+
+async function cleanup(guild) {
+  const categoryPrivateChat = guild.channels.cache.find(
+    ({name, type}) =>
+      type === 'category' && name.toLowerCase().includes('private chat'),
```
Comment (kentcdodds) [inline src/private-chat/cleanup.js:6]: To be clear, I'll need to make a category titled "Private Chat" for this to work right?

File: src/private-chat/cleanup.js (hunk 3 above)
Comment (kentcdodds) [reply src/private-chat/cleanup.js:11]: I'm not worried about this. I doubt we'll run into limits. If we do then we can lower it to 5 minutes.

**Review by kentcdodds — APPROVED** [review]
I love this. It's going to be really helpful!

File: src/commands/command-fns/private-chat.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+const {privateChannelPrefix} = require('../utils')
+
+async function privateChat(message) {
+  const mentionedMembers = Array.from(message.mentions.members.values())
+  const mentionedMembersNicknames = Array.from(
+    message.mentions.members.values(),
+  ).map(m => m.nickname ?? m.user.username)
+  if (mentionedMembers.length < 2) {
+    return message.channel.send(`You should mention at least two members.`)
+  }
```
Comment (kentcdodds) [inline src/commands/command-fns/private-chat.js]: Actually, we should only require mentioning 1 name and we should automatically include the requestor in the room as well.

Comment (kentcdodds) [reply src/commands/command-fns/private-chat.js]:
```suggestion
  const mentionedMembers = [...Array.from(message.mentions.members.values()), message.member]
  const mentionedMembersNicknames = Array.from(
    message.mentions.members.values(),
  ).map(m => m.nickname ?? m.user.username)
  if (mentionedMembers.length < 2) {
    return message.channel.send(`You should mention at least one other member.`)
  }
```

**Comment by kentcdodds** [comment]
Made a few small adjustments in 5d0faaa

**Comment by marcosvega91** [comment]
Sorry I was sleeping hahhahahh. Thanks for the changes ☺️

**Pull request merged**